src/test/java/seedu/address/logic/commandhistory/SuppliedCommandHistorySelectorTest.java

```diff
@@ -0,0 +1,110 @@
+        typicalSelector.navigateToOnePastLast();
+        assertEquals(Optional.empty(), typicalSelector.selectNextUntilOnePastLast());
+    }
+
+    @Test
+    public void navigateToOnePastLast_thenSelectPreviousOnNonEmpty_returnsLastEntry() {
+        String lastEntry = typicalHistory.get(typicalHistory.size() - 1).value;
+
+        typicalSelector.navigateToOnePastLast();
+        assertEquals(Optional.of(lastEntry), typicalSelector.selectPreviousUntilFirst());
+    }
+
+    @Test
+    public void selectNext_onEmpty_returnsEmpty() {
+        assertEquals(Optional.empty(), emptySelector.selectNextUntilOnePastLast());
+    }
+
+    @Test
+    public void selectNext_onNonEmptyFromLast_returnsEmpty() {
+        List<CommandHistoryEntry> entries = getEntries(typicalHistory);
+        // Navigate to last
+        typicalSelector.navigateToOnePastLast();
+        typicalSelector.selectPreviousUntilFirst();
+
+        assertEquals(Optional.empty(), typicalSelector.selectNextUntilOnePastLast());
+    }
+
+    @Test
+    public void selectNext_repeatedOnNonEmptyFromFirst_canSelectAllEntries() {
+        List<CommandHistoryEntry> entries = getEntries(typicalHistory);
+        typicalSelector.navigateToOnePastLast();
+        // Navigate to first entry
+        for (int i = entries.size() - 1; i >= 0; i--) {
+            typicalSelector.selectPreviousUntilFirst();
+        }
+
+        // Check each entry from second until last
+        for (int i = 1; i < entries.size(); i++) {
+            Optional<String> entryText = Optional.of(entries.get(i).value);
+            assertEquals(entryText, typicalSelector.selectNextUntilOnePastLast());
+        }
+    }
+
+    @Test
+    public void selectPrevious_onEmpty_returnsEmpty() {
+        assertEquals(Optional.empty(), emptySelector.selectPreviousUntilFirst());
+    }
+
+    @Test
+    public void selectPrevious_onNonEmptyFromFirst_returnsFirst() {
+        List<CommandHistoryEntry> entries = getEntries(typicalHistory);
+        typicalSelector.navigateToOnePastLast();
+        // Navigate to first entry
+        for (int i = entries.size() - 1; i >= 0; i--) {
+            typicalSelector.selectPreviousUntilFirst();
+        }
+
+        // After selecting all the way to the first, the next call should return the first entry.
+        assertEquals(Optional.of(entries.get(0).value), typicalSelector.selectPreviousUntilFirst());
+    }
+
+    @Test
+    public void selectPrevious_repeatedOnNonEmptyFromLast_canSelectAllEntries() {
+        List<CommandHistoryEntry> entries = getEntries(typicalHistory);
+
+        // Check all entries in reverse
+        typicalSelector.navigateToOnePastLast();
+        for (int i = entries.size() - 1; i >= 0; i--) {
+            Optional<String> entryText = Optional.of(entries.get(i).value);
+            assertEquals(entryText, typicalSelector.selectPreviousUntilFirst());
+        }
+    }
+
+    @BeforeEach
```

Comment:
I think this should be at the top of the file to be clear what happens

Reply:
Ah good point. Thanks.